### PR TITLE
Feature/add LSM programs support

### LINKED ISSUE #1686: LSM Support Skeleton
## Prerequisites

- [x] Is this an EPIC issue describing other issues ? (add label: EPIC)

## Feature description

- Load the Program
- Block something
- Alert through tracee-rules

## Context

image

## Additional Information

Being discussed between @yanivagman and @rafaeldtinoco

### PR DESCRIPTION
**Pull request by AlonZivony**
### 1. Explain what the PR does

Add LSM program support.
The LSM program will only be loaded in systems that support it.
In other systems the event using it will fail - as expected of it.
This PR is also fixing an edge case check needed for LSM programs support that was missing.

Resolves #1686

### 2. Explain how to test it

Added both unit and integration tests

### CODE REVIEW

File: common/environment/lsm.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+// CheckLSMSupport checks if BPF LSM is supported using runtime securityfs detection with kernel config fallback.
+// It first attempts runtime detection, falling back to kernel config if securityfs is not mounted.
+func CheckLSMSupport(filesystem fs.FS, getKernelConfigValue KernelConfigValueFunc) (bool, error) {
+	// Check runtime LSM enablement using the provider's filesystem
+	runtimeEnabled, err := IsLSMSupportedInSecurityFs(filesystem)
+	if err != nil {
+		// If security fs is not mounted, fall back to kernel config check
+		if errors.Is(err, ErrSecurityFSNotMounted) {
+			return CheckLSMSupportInKconfig(getKernelConfigValue)
```
Comment (yanivagman) [inline common/environment/lsm.go]: Let's assume LSM fs is not mounted but kernel was configured with LSM BPF support - this can lead to false positive in the case LSM was not enabled during boot, can't it?

File: go.mod [hunk 2]
```diff
@@ -170,3 +170,5 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	kernel.org/pub/linux/libs/security/libcap/psx v1.2.76 // indirect
 )
+
+replace github.com/aquasecurity/tracee/common => ./common
```
Comment (yanivagman) [inline go.mod]: Instead of replacing common to local, this PR should be split to two PRs - one that update common package, and once merged, another PR that uses it

File: pkg/ebpf/environment_provider.go
Comment (yanivagman) [inline pkg/ebpf/environment_provider.go:1]: For maintainability, this environment provider should probably be in its own subpackage and not directly under the ebpf package.

I also think that it should be placed in the same package where the environment provider interface is placed, which is currently under the probes package

File: pkg/ebpf/probes/lsm.go [hunk 3]
```diff
@@ -3,7 +3,7 @@ package probes
 import (
 	bpf "github.com/aquasecurity/libbpfgo"
 
-	"github.com/aquasecurity/tracee/pkg/errfmt"
+	"github.com/aquasecurity/tracee/common/errfmt"
```
Comment (yanivagman) [inline pkg/ebpf/probes/lsm.go:6]: This should be amended to the commit that introduced it so it will compile correctly

File: tests/integration/lsm_test.go [hunk 4]
```diff
@@ -2,17 +2,19 @@ package integration
...
 	"testing"
 	"time"
 
-	bpf "github.com/aquasecurity/libbpfgo"
 	"go.uber.org/goleak"
 
+	bpf "github.com/aquasecurity/libbpfgo"
+
+	"github.com/aquasecurity/tracee/common/environment"
+	"github.com/aquasecurity/tracee/common/logger"
```
Comment (yanivagman) [inline tests/integration/lsm_test.go]: ditto

**Review by yanivagman — CHANGES_REQUESTED** [review]

File: pkg/ebpf/environment_provider.go
Comment (AlonZivony) [reply pkg/ebpf/environment_provider.go:1]: I do not agree with you about the subpackage outside of the ebpf package as this is an interface created specifically for the compatibility checks. It doesn't have any reason for existing outside the context of its usage.
In the end I used the common Golang practice of creating an interface for imported types to allow mocking and as a result testability. This is why I agree with putting the implementation in the same package. But moving the interface outside will miss the practice point and will make it a standalone interface that the package has to fit itself to it instead of the other way around.

File: go.mod (hunk 2 above)
Comment (AlonZivony) [reply go.mod]: Made a separate PR - #4914

File: pkg/ebpf/environment_provider.go
Comment (yanivagman) [reply pkg/ebpf/environment_provider.go:1]: I think you misunderstood what I said.
I wrote it should not be **DIRECTLY** under ebpf package, but in its own subpackage.
ebpf/probes/environment_provider.go, for example, will not be directly under ebpf package but still in the context of its usage

File: go.mod (hunk 2 above)
Comment (yanivagman) [reply go.mod]: Merged #4914, you can now rebase your work on top of it

**Comment by AlonZivony** [comment]
Currently the tests are failing because of a thing I have missed with the LSM compatibility check - if the lsm file is missing from the securityfs dir, it means that it is not supported. Right now it creates an error for it which futile as it is expected in some systems (as it is in the tests).
I will update the `common/environment` package to do this check and update this PR with it.

**Comment by AlonZivony** [comment]
Created a PR for fixing the `common/environment` package - #4919

File: pkg/events/core.go [hunk 5]
```diff
@@ -13888,4 +13889,18 @@ var CoreEvents = map[ID]Definition{
 			ids: []ID{ExecTest},
 		},
 	},
+	LsmTest: {
```
Comment (yanivagman) [inline pkg/events/core.go:15192]: missing internal event in the definition

File: tests/integration/lsm_test.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	for _, p := range policies {
+		initialPolicies = append(initialPolicies, p)
+	}
+	testConfig.InitialPolicies = initialPolicies
+
+	return testConfig, nil
+}
+
+// isLsmSupported checks if the current kernel supports LSM BPF programs
+func isLsmSupported() (bool, error) {
```
Comment (yanivagman) [inline tests/integration/lsm_test.go]: dup of common/environment

**Comment by yanivagman** [comment]
> Created a PR for fixing the `common/environment` package - #4919

Merged. Please update common module.

File: pkg/events/core.go (hunk 5 above)
Comment (AlonZivony) [reply pkg/events/core.go:15192]: So after looking at it, the event is under the test events which are not defined right now as internal events.
They are automatically omitted from the print utilities of events.

**Comment by AlonZivony** [comment]
Changed to draft for debugging the issue with AMZ2 aarch64

File: tests/e2e-inst-test.sh [hunk 7]
```diff
@@ -287,7 +318,7 @@ for TEST in $TESTS; do
         anyerror="${anyerror}$TEST,"
 
         info "$TEST: FAILED, stderr from tracee:"
-        cat $logfile
+        grep -E 'L":"(ERROR|WARN|FATAL)' "$logfile"
```
Comment (yanivagman) [inline tests/e2e-inst-test.sh:321]: I don't think this change is related to this PR

File: pkg/ebpf/probes/cgroup.go [hunk 8]
```diff
@@ -111,11 +110,7 @@ func (p *CgroupProbe) detach(args ...interface{}) error {
 	// kernels) might not be done using BpfLink logic. Without a file
 	// descriptor for the link, tracee needs to raise its capabilities
 	// in order to call bpf() syscall for the legacy detachment.
-	err = capabilities.GetInstance().EBPF(
-		func() error {
-			return p.bpfLink.Destroy()
-		},
-	)
+	err = p.bpfLink.Destroy()
```
Comment (yanivagman) [inline pkg/ebpf/probes/cgroup.go:113]: Will that be a problem (removing the capability)?

File: tests/e2e-inst-test.sh [hunk 9]
```diff
@@ -160,6 +172,7 @@ tracee_command="./dist/tracee \
                     --output option:sort-events \
                     --output option:parse-arguments \
                     --output json:$outputfile \
+                    --log debug \
```
Comment (yanivagman) [inline tests/e2e-inst-test.sh:175]: This should be removed

File: tests/e2e-inst-test.sh [hunk 10]
```diff
@@ -196,7 +209,7 @@ if [[ $timedout -eq 1 ]]; then
     info "$TEST: timed out"
     info "$TEST: FAILED. ERRORS:"
     info
-    cat $logfile
+    grep -E 'L":"(ERROR|WARN|FATAL)' "$logfile"
```
Comment (yanivagman) [inline tests/e2e-inst-test.sh:212]: This should be removed

File: tests/e2e-inst-test.sh (hunk 7 above)
Comment (AlonZivony) [reply tests/e2e-inst-test.sh:321]: Actually it is.
To be able to check if the logs I expect to be printed about cancelling the event has been printed, I have to get all of the logs. This is why I have added the grep with the WARN|ERROR|FATAL

File: pkg/ebpf/probes/cgroup.go (hunk 8 above)
Comment (AlonZivony) [reply pkg/ebpf/probes/cgroup.go:113]: Shouldn't be, as it is called from inside the `initBPF` function, which is called with the EBPF ring capabilities:
```Go
err = capabilities.GetInstance().EBPF(
    func() error {
	    return t.initBPF()
    },
)
```
The only danger here is that in the future for some reason the dependencies manager will call it because of some changes during runtime without an active eBPF capabilities usage. But for now it is safe.
We will need a long term solution for the capabilities, like the reenter lock.

File: tests/e2e-inst-test.sh (hunk 9 above)
Comment (AlonZivony) [reply tests/e2e-inst-test.sh:175]: ditto

File: tests/e2e-inst-test.sh (hunk 10 above)
Comment (AlonZivony) [reply tests/e2e-inst-test.sh:212]: ditto
